### PR TITLE
Allow user to track unlimited blogs

### LINKED ISSUE #297: Blog tracker: track unlimited blogs
I don't know if this is the place to put it, but I sent an ask, but then realized the Tumblr is probably swamped, and thought maybe here was a better place to put this. (I can type more here, anyway.)

Anyway, I was wondering if there is any way we could make a new extension, or modify, the current blog tracker extension. (Which seems to be broken for me at the moment.) First of all, it would be nice if you could bookmark more than 10 blogs. (Xkit guy said this wasn't possible because the post numbers would break, but I don't care about the posts numbers.) Secondly, when you would click on the blog it could either open in View On Dash (Please no, it has never worked with photosets.) or ~~the default Tumblr sidebar-for-blogs-you-click-on-thing-that-pops-up-on-the-right-idk-what-to-call-it.~~ in peepr. This way you could closely follow more than 10 blogs, and easily view them without leaving your dash.

Reasons I want this extension, because some of you may be thinking "Just unfollow the blogs you don't want to see then!!" I want to see all blogs that I follow, but I also follow about 400... So it isn't practical for me to see every single post on my dash. I also have a lot of friends I have made on Tumblr, and this number is over 10. My friends and I tend to talk to each other in tags, so it has become a habit of mine to check their blogs daily before I venture out into the wilds known as my dash. Having this extension would allow me to quickly click, view in ~~sidebar-thing~~ peepr, and check for tags, and then be on my merry way.

Anyway, thank you for reading. Now I shall go and keep my fingers crossed. Ahahahahah.

### PR DESCRIPTION
**Pull request by bvtsang**
In anticipation of people adding many more tracked blogs, I've introduced a delay in calling Tumblr's API. I'm not sure if this delay needs to be adjusted.

For this pull request, I'm particularly interested in feedback related to performance and other issues that tracking unlimited blogs might cause.

Fixes #297. Includes fixes I had originally planned for #332.

### CODE REVIEW

File: Extensions/people_notifier.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 //* TITLE Blog Tracker **//
-//* VERSION 0.4.1 **//
+//* VERSION 0.5.0 **//
 //* DESCRIPTION Track people like tags **//
 //* DEVELOPER STUDIOXENIX **//
-//* DETAILS Blog Tracker lets you track blogs like you can track tags. Add them on your dashboard, and it will let you know how many new posts they've made the last time you've checked their blogs, or if they've changed their URLs. **//
+//* DETAILS Blog Tracker lets you track blogs like you can track tags. Add them on your dashboard, and it will let you know how many new posts they've made the last time you've checked their blogs, or if they've changed their URLs.<br><br>Please be aware that the more blogs you add, the longer it will take to track them all. **//
```
Comment (blackjackkent) [inline Extensions/people_notifier.js]: Do we want to bold or emphasize this as a warning somehow? (Also maybe something like "Please be aware that adding more than ten blogs may cause the tracker extension to load slowly" so they have a sense of how far they can go before they have to start being concerned?)

**Comment by blackjackkent** [comment]
This LGTM...added one comment regarding the extension description.

**Comment by nightpool** [comment]
maybe we should add a one-time warning if you try to add more then 10 blogs?

**Comment by nightpool** [comment]
also wait why would this slow the page down? shouldn't it be async?

**Comment by blackjackkent** [comment]
Hence why I said "may cause the extension to load slowly." The issue isn't that it would load the page slowly, but all your tracked blogs might take a while to come up. (at least that's my understanding of the situation anyway)

**Comment by nightpool** [comment]
thoughts—a "check tracked blogs" button? do we even really need to show
these numbers immediately on pageload?

File: Extensions/people_notifier.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
 //* TITLE Blog Tracker **//
-//* VERSION 0.4.1 **//
+//* VERSION 0.5.0 **//
```
Comment (nightpool) [inline Extensions/people_notifier.js]: probably can 1.0 this. i feel fine with how stable it is

**Comment by blackjackkent** [comment]
For my part I use this extension regularly specifically because it notifies me of certain people I want to keep an eye on without me having to actively retrieve their blog info, so I think that's a strong part of its appeal.

**Comment by nightpool** [comment]
changed my mind on the version thing, 0.5 is probably fine.

my only outstanding concern is that someone will (on accident or maliciously) add 50+ blogs, and cause tumblr to ban our api key. Do they guidelines on rate limiting we could follow?

@blackjackkent fair

**Comment by nightpool** [comment]
maybe we should just raise the limit instead of removing it all together? to 30 or something?

File: Extensions/people_notifier.js (hunk 1 above)
Comment (bvtsang) [inline Extensions/people_notifier.js]: My impression is that extension details aren't commonly read, so bolding this warning might not catch too much attention. However, what I could do is also include this warning in the window that pops up when adding a new blog. I think users are more likely to see that than read the extension details.

**Comment by bvtsang** [comment]
@nightpool I share your concern about Tumblr banning our API key. Here's everything I know about this extension and rate limiting:

- Blog Tracker calls Tumblr's [`/posts` API method](https://www.tumblr.com/docs/en/api/v2#posts)
- Blog Tracker will only perform an API call to check a blog if it hasn't checked within 10 minutes
- As of this pull request, I'm adding an API call delay of 0.25s per added blog
- Although one anecdote says that Tumblr's rate limit is [1000 requests/minute](http://stackoverflow.com/a/27875468/2073440 "Tumblr API call or request limits - Stack Overflow"), its true rate limit is a mystery
- The previous point is a partial lie: there are apparently [headers](https://stackoverflow.com/questions/24854293/tumblr-api-v2-error-code-429-rate-limit-exceeded-you-are-being-rate-limited) that tell hint at what the limits are

Things I need to do:

- [ ] Add a warning to popup that shows when adding a new blog to track
- [ ] Investigate the headers that hint at the API limits

**Comment by bvtsang** [comment]
Using [Tumblr's Python client](https://github.com/tumblr/pytumblr), I inserted a print statement in their `request.py`:

```python
def get(self, url, params):
    """
    Issues a GET request against the API, properly formatting the params

    :param url: a string, the url you are requesting
    :param params: a dict, the key-value of all the paramaters needed
                   in the request
    :returns: a dict parsed of the JSON response
    """
    url = self.host + url
    if params:
        url = url + "?" + urllib.urlencode(params)

    client = oauth.Client(self.consumer, self.token)
    try:
        client.follow_redirects = False
        resp, content = client.request(url, method="GET", redirections=False, headers=self.headers)
    except RedirectLimit, e:
        resp, content = e.args

    print resp
    return self.json_parse(content)
```

And then tried

```python
import pytumblr

client = pytumblr.TumblrRestClient(
	'<consumer_key>',
	'<consumer_secret>',
	'<oauth_token>',
	'<oauth_secret>',
)

print "Started: " + str(datetime.datetime.utcnow())

for i in range(0, 500):
	print str(i) + ":\n"
	client.posts('xumbra')
	print "\n\n"

print "Ended: " + str(datetime.datetime.utcnow())
```

and got the following headers:

```json
{
    'status':'200',
    'content-length':'118628',
    'content-location':    u'https://api.tumblr.com/v2/blog/xumbra.tumblr.com/posts?oauth_body_hash=2jmj7l5rSw0yVb%2FvlWAYkK%2FYBwk%3D&oauth_nonce=30526426&oauth_timestamp=1440780831&oauth_consumer_key=<consumer_key>&oauth_signature_method=HMAC-SHA1&oauth_version=1.0&oauth_token=<oauth_token>&oauth_signature=<oauth_signature>&api_key=<api_key>',
    'accept-ranges':'bytes',
    'vary':'Accept-Encoding',
    'server':'nginx',
    'connection':'keep-alive',
    '-content-encoding':'gzip',
    'x-ua-compatible':'IE=Edge,chrome=1',
    'date':'Fri, 28 Aug 2015 16:53:54 GMT',
    'p3p':'CP="Tumblr\'s privacy policy is available here: https://www.tumblr.com/policy/en/privacy"',
    'content-type':'application/json; charset=utf-8'
}
```

with the start and stop times:

```
Started: 2015-08-28 17:19:33.298000
Ended: 2015-08-28 17:26:30.708000
```

That's about 7 minutes, so 500 requests / 7 minutes = 71.4 requests/minute. And I see no sign of the [anything in the headers that hints at API limits](https://groups.google.com/forum/#!topic/tumblr-api/RJzLT3YsiYI).

My immediate thoughts after seeing these headers:

- Maybe the results are being cached server-side, but I don't know if that affects anything
- Maybe the `x_ratelimit_api_*` headers show up only when I've hit the limit?
- The Python client has no internal throttling (I checked)
- My computer probably isn't making requests fast enough, at least alone

What do you all think?

**Comment by bvtsang** [comment]
Since no one provided feedback on my rate limiting research, I've implemented @nightpool's suggestion and raised the limit to 30.

/CC @blackjackkent @nightpool for previously reviewing

File: Extensions/people_notifier.js [hunk 3]
```diff
@@ -79,7 +79,22 @@ XKit.extensions.people_notifier = new Object({
 
 	},
 
+	should_render: function() {
```
Comment (nightpool) [inline Extensions/people_notifier.js]: wasn't this fixed in another PR, possibly by @blackjackkent?

**Comment by nightpool** [comment]
Other then the comment about should_render, LGTM.

squash commits?

File: Extensions/people_notifier.js (hunk 3 above)
Comment (rosalindwills) [inline Extensions/people_notifier.js]: Yeah, I did something related down on line 348...

Comment (rosalindwills) [inline Extensions/people_notifier.js]: oh ffs

Comment (nightpool) [inline Extensions/people_notifier.js]: Nevermind, I was thinking of https://github.com/new-xkit/XKit/pull/398

**Comment by nightpool** [comment]
said in chat but repeating here for posterity. LGTM as long as we get it squashed and @blackjackkent has reviewed it for duplicate additions with her changes.

**Comment by bvtsang** [comment]
@nightpool Squashed and changed the commit message.

**Comment by nightpool** [comment]
LGTM (and checked with @blackjackkent about URL stuff)

**Comment by nightpool** [comment]
@homu r+

**Pull request merged**